Selector Stats documentation (#2399)
MicrosoftDocs/edge-developer: microsoft-edge/devtools-guide-chromium/evaluate-performance/selector-stats.md

> ---

Reviewer: @zoherghadyali - add entry in TOC.yml

https://github.com/MicrosoftDocs/edge-developer/blob/user/zoghadya/selector-stats/microsoft-edge/toc.yml

Reviewer: this was done so can Resolve this comment thread

title: Analyze selector performance during Recalculate Style events
description: Use the Selector Stats feature in the Performance tool to understand which of your CSS selectors are taking up the most time during Recalculate Style events.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.prod: microsoft-edge
ms.date: 01/09/2023
---
# Analyze selector performance during Recalculate Style events

> The **Performance** tool highlights long tasks with red triangles to indicate work on the main thread that takes too long and has slow performance.  

Reviewer: ```suggestion
The **Performance** tool highlights long tasks with red triangles to indicate work on the main thread that takes too long and has slow performance:
```
opportunity to use colon to glue flow (equiv of figure captions & references, but low-key)

> ![Long Task indicators for events on the main thread with slow performance.](../media/evaluate-performance-performance-animation-frame-fired.msft.png)

Reviewer: ```suggestion
![Long Task indicators for events on the main thread with slow performance](../media/evaluate-performance-performance-animation-frame-fired.msft.png)
```
moving away from period, this might be noted in wiki > Edge Dev Style Guide.  for simple-as-possible Markdown

> ![Long Task indicators for events on the main thread with slow performance.](../media/evaluate-performance-performance-animation-frame-fired.msft.png)

Reviewer: Can red boxes be added?  No idea what to look at in this capture.  Where am I supposed to be looking?

GH rendered:
https://github.com/MicrosoftDocs/edge-developer/blob/user/zoghadya/selector-stats/microsoft-edge/devtools-guide-chromium/evaluate-performance/selector-stats.md#analyze-selector-performance-during-recalculate-style-events

Reviewer 2: I would also suggest creating a new image instead of reusing an existing one.
Our current way of dealing with images is that each md file should really have its own image folder with its own images. 
Reusing images across articles usually leads to problems down the line, even if it feels like we're saving time and cutting on file numbers initially.

Reviewer: caution, Outdated comment thread.
"its own image folder", that would be:
microsoft-edge/devtools-guide-chromium/evaluate-performance/selector-stats-images/

Reviewer: @captainbrosset - I entered a fresh non-Outdated [comment](https://github.com/MicrosoftDocs/edge-developer/pull/2399/files#r1068561465) to create a dedicated images dir.

Reviewer 3: > I would also suggest creating a new image instead of reusing an existing one. Our current way of dealing with images is that each md file should really have its own image folder with its own images. Reusing images across articles usually leads to problems down the line, even if it feels like we're saving time and cutting on file numbers initially.

fixed

> In your performance recordings, some of these long-running tasks may be **Recalculate Style** events.  These events track the time it takes for the browser to iterate through DOM elements on a page, find all of the CSS style rules that match a given element, and then compute the element's actual style based on these rules.  These styles need to be recalculated whenever the applicabiliy of CSS rules may have changed, like when:

Reviewer: ```suggestion
In your performance recordings, some of these long-running tasks may be **Recalculate Style** events.  These events track the time it takes for the browser to iterate through DOM elements on a page, find all of the CSS style rules that match a given element, and then compute the element's actual style based on these rules.  These styles need to be recalculated whenever the applicabiliy of CSS rules may have changed, such as when:
```

> - Elements are added or removed from the DOM
> - An element's attributes, like `class` or `id`, were changed
> - User input, like mouse move or focus change, which can affect `:hover` rules

Reviewer: ```suggestion
- Elements are added or removed from the DOM.
- An element's attributes, such as `class` or `id`, were changed.
- User input occurred, such as a mouse move or focus change, which can affect `:hover` rules.
```

When you have long running **Recalculate Style** events, you can use the **Selector Stats** feature to understand which of your CSS selectors are taking up the most time and resulting in the slow performance.

## How to use Selector Stats

> To see **Selector Stats** during **Recalculate Style** events, complete the following steps.

Reviewer: ```suggestion
To see **Selector Stats** during **Recalculate Style** events:
```
streamlined

> 1. In the **Performance** tool, click **Capture settings** (![Capture settings.](../media/capture-settings-icon.msft.png)).

Reviewer: ```suggestion
1. In the **Performance** tool, click **Capture settings** (![Capture settings](../media/capture-settings-icon.msft.png)).
```

> 1. Click **Enable advanced rendering instrumentation (slow)** to turn the setting on. Note that this setting adds more overhead for performance recordings so you should only leave it turned on if you need to record **Selector Stats** and other rendering information.

Reviewer: ```suggestion
1. Click **Enable advanced rendering instrumentation (slow)** to turn the setting on. Note that this setting adds more overhead for performance recordings, so you should only leave it turned on when you need to record **Selector Stats** and other rendering information.
```

> 1. Click **Record** and run the scenario you want to improve for your website or app.

Reviewer: ```suggestion
1. Click **Record** and run the scenario that you want to improve for your website or app.
```
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/t/that-vs-which - "**Global tip:** Include _that_ even if the sentence is clear without it. It helps to clarify the sentence structure."

1. Click **Stop**.

> 1. Now find a **Recalculate Style** event. In the bottom section of the **Performance** tool, select **Selector Stats**.

Reviewer: ```suggestion
1. Now find a **Recalculate Style** event. In the bottom section of the **Performance** tool, select **Selector Stats**:
```
opportunity to use colon to glue flow

Reviewer 2: We also need to tell the reader to click on the Recalculate Style event, not just to find it.

>    ![Selector Stats in the Performance tool.](images/performance-tool-selector-stats.png)

Reviewer: ```suggestion
   ![Selector Stats in the Performance tool](images/performance-tool-selector-stats.png)
```

> The **Selector Stats** section of the **Performance** tool shows you a table of selectors, the amount of time the browser spent calculating the style rules, and the number of elements the selector matched.

Reviewer: ```suggestion
The **Selector Stats** section of the **Performance** tool shows you:
* A table of selectors.
* The amount of time the browser spent calculating the style rules.
* The number of elements the selector matched.
```
opportunity to convert in-sentence list to explicit list.  Acrolinx might suggest this.  One advantage of list format is to avoid formatting all content as prose block paragraphs.

Reviewer 2: Let's replace "A table of selectors" with "A table of CSS selectors", just to be extra clear.

## Analyze Selector Stats

> You can sort the data in the **Selector Stats** table in ascending or descending order for any of the columns by clicking the column header.  For example, sort the **Elapsed (µs)** column in descending order to see which selectors take up the most time in the **Recalculate Style** event you have selected.

Reviewer: Easy comment to resolve:  Pls check whether µs (microseconds) renders at review.lmc:
https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/evaluate-performance/selector-stats?branch=pr-en-us-2399

It successfully renders in GitHub's view:
https://github.com/MicrosoftDocs/edge-developer/blob/user/zoghadya/selector-stats/microsoft-edge/devtools-guide-chromium/evaluate-performance/selector-stats.md#analyze-selector-stats

Reviewer 2: It does render fine with our font. I would tend to think that this glyph is common enough that most fonts have it.

> As a general rule, if you find a selector that took the browser a long time to calculate and was matched many times, that would be a good candidate to try and improve.  Try simplifying the selector or making it more specific so that fewer elements match to it.  As a result of making these improvements, repeat the steps from the previous section, and you should observe the time the browser spends in the **Recalculate Style** event decrease.

Reviewer: ```suggestion
As a general rule, if you find a selector that took the browser a long time to calculate and was matched many times, that would be a good candidate to try and improve.  Try simplifying the selector or making it more specific so that fewer elements match to it.  As a result of making these improvements, repeat the steps from the previous section, and you should observe that the time the browser spends in the **Recalculate Style** event decrease.
```
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/t/that-vs-which - 
"Include _that_ even if the sentence is clear without it. It helps to clarify the sentence structure."

Reviewer 2: I'm unclear whether "making it more specific" is the right thing to say here. People might interpret it as: "Oh I need to add more parents to my selector for it to be more specific", which might also cause the browser to do more work since it has to go up the chain of ancestors and check more things that it should.

Saying how to improve selectors is dangerous business, and I think our docs should stay away from this as much as we can. We provide a tool, we tell people to measure before and after, and that's great. I don't think we should necessarily tell them how to fix their selectors because that will highly depend on their particular case.

```suggestion
As a general rule, if you find a CSS selector that took the browser a long time to calculate and was matched many times, that's a good candidate to try and improve.  Try to change your selector so it requires less time to calculate and matches fewer elements on the page.  How to improve your CSS selectors depend on your particular use case. Repeat the steps from the previous section to confirm that your changes helped decrease the **Recalculate Style** event duration.
```

<!-- ====================================================================== -->
## See also

> * [Get started analyzing runtime performance](index.md) - An introduction to analyzing runtime performance in DevTools.

Reviewer: ```suggestion
* [Get started analyzing runtime performance](./index.md) - An introduction to analyzing runtime performance in DevTools.
```
per autocomplete & default pathing in VS Code w/ doc'n extensions

> * [Performance features reference](reference.md), - A reference for many of the features offered in the Performance tool.

Reviewer: ```suggestion
* [Performance features reference](./reference.md) - A reference for many of the features offered in the **Performance** tool.
```

---
title: Analyze selector performance during Recalculate Style events
description: Use the Selector Stats feature in the Performance tool to understand which of your CSS selectors are taking up the most time during Recalculate Style events.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.prod: microsoft-edge
ms.date: 01/09/2023
---
# Analyze selector performance during Recalculate Style events

The **Performance** tool highlights long tasks with red triangles to indicate work on the main thread that takes too long and has slow performance:

> ![Long Task indicators for events on the main thread with slow performance](../media/evaluate-performance-performance-animation-frame-fired.msft.png)

Reviewer: Move png to new, per-.md dir:
microsoft-edge/devtools-guide-chromium/evaluate-performance/**selector-stats-images**/

Reviewer: done, see latest Files Changed https://github.com/MicrosoftDocs/edge-developer/pull/2399/files

Pls resolve this thread

> ![Long Task indicators for events on the main thread with slow performance](../media/evaluate-performance-performance-animation-frame-fired.msft.png)

Reviewer: redo screenshot w/o Focus Mode

Reviewer: done, GT rendered
https://github.com/MicrosoftDocs/edge-developer/blob/user/zoghadya/selector-stats/microsoft-edge/devtools-guide-chromium/evaluate-performance/selector-stats.md

Pls Resolve this thread

> ![Long Task indicators for events on the main thread with slow performance](./selector-stats-images/performance-tool-selector-stats.png)selector-stats-images

Reviewer: ```suggestion
![Long Task indicators for events on the main thread with slow performance](./selector-stats-images/performance-tool-selector-stats.png)
```

In your performance recordings, some of these long-running tasks may be **Recalculate Style** events.  These events track the time it takes for the browser to iterate through DOM elements on a page, find all of the CSS style rules that match a given element, and then compute the element's actual style based on these rules.  These styles need to be recalculated whenever the applicability of CSS rules may have changed, such as when:

- Elements are added to or removed from the DOM.
- An element's attributes are changed, such as `class` or `id`.
- User input occurs, such as a mouse move or focus change, which can affect `:hover` rules.

> When you have long running **Recalculate Style** events, you can use the **Selector Stats** feature to understand which of your CSS selectors are taking up the most time and resulting in slow performance.

Reviewer: ```suggestion
When you have long-running **Recalculate Style** events, you can use the **Selector Stats** feature to understand which of your CSS selectors are taking up the most time and resulting in slow performance.
```

## How to use Selector Stats

To see **Selector Stats** during **Recalculate Style** events:

1. In the **Performance** tool, click **Capture settings** (![Capture settings icon](../media/capture-settings-icon.msft.png)).

1. Click **Enable advanced rendering instrumentation (slow)** to turn the setting on. Note that this setting adds more overhead for performance recordings, so you should only leave it turned on when you need to record **Selector Stats** and other rendering information.
 
1. Click **Record** and run the scenario that you want to improve for your website or app.

1. Click **Stop**.

> 1. Now find a **Recalculate Style** event and click it.

Reviewer: ```suggestion
1. Find a **Recalculate Style** event and click it.
```

1. In the bottom section of the **Performance** tool, click **Selector Stats**:

   ![Selector Stats in the Performance tool](./selector-stats-images/performance-tool-selector-stats.png)
 
The **Selector Stats** section of the **Performance** tool shows you:
* A table of CSS selectors.
* The amount of time the browser spent calculating the style rules.
* The number of elements the selector matched.

> ## Analyze Selector Stats

Reviewer: ```suggestion

<!-- ====================================================================== -->
## Analyze Selector Stats
```
